### ISSUE TITLE
Defined layers but received error:  __init__() missing 1 required positional argument: 'layer'

### ISSUE THREAD

**Issue by bavanhassel** [tensorflow/tensorflow#43771]
print(tf.__version__) 2.2.0
print(keras.__version__) 2.3.0-tf

Message:
 __init__() missing 1 required positional argument: 'layer'

I would like to better understand why I get this message with the following definition used in the example:

https://aihub.cloud.google.com/p/products%2F2290fc65-0041-4c87-a898-0289f59aa8ba

Model function definition:

def model_fn(features, labels, mode, params = params):
    # import the data and unpack the features
    # serving input_fn returns a dict, convert to multivalue obj
    if isinstance(features, dict):
        features = features['words'], features['length']
   
    words, length = features
   
    # Embedding
    embedding = tf.Variable(tf.random.normal([params['max_words'], params['dim']]))
    embedding_lookup_for_x = tf.nn.embedding_lookup(embedding, words)
   
    # LSTM
    lstm_cell_fw = tf.keras.layers.LSTMCell(params['lstm_size'])
    lstm_cell_bw = tf.keras.layers.LSTMCell(params['lstm_size'])
    states, final_state = tf.keras.layers.Bidirectional(
                                        cell_fw = lstm_cell_fw,
                                        cell_bw = lstm_cell_bw,
                                        inputs = embedding_lookup_for_x,
                                        dtype = tf.float32,
                                        time_major = False,
                                        sequence_length = length)
    lstm_out = tf.keras.layers.concat([states[0], states[1]], axis = 2)

The original definition was (no longer supported):

    # LSTM
#    lstm_cell_fw = tf.nn.rnn_cell.BasicLSTMCell(params['lstm_size'], state_is_tuple = True)
#    lstm_cell_bw = tf.nn.rnn_cell.BasicLSTMCell(params['lstm_size'], state_is_tuple = True)
#    states, final_state = tf.nn.bidirectional_dynamic_rnn(
#                                        cell_fw = lstm_cell_fw,
#                                        cell_bw = lstm_cell_bw,
#                                        inputs = embedding_lookup_for_x,
#                                        dtype = tf.float32,
#                                        time_major = False,
#                                        sequence_length = length)
#    lstm_out = tf.concat([states[0], states[1]], axis = 2)

Please advise.

**Comment by Saduf2019** [comment]
@bavanhassel 
Please share complete stand alone indented code for us to replicate the issue faced or if possible share a colab gist with the error.
With respect tot he error faced please refer to these issues and let us know:
[link](https://stackoverflow.com/questions/56106546/typeerror-init-missing-1-required-positional-argument-units), [link1](https://github.com/prodo56/Time-series-analysis-using-RNN/issues/1#issuecomment-325418816)

**Comment by bavanhassel** [comment]
import tensorflow as tf
from functools import partial
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.utils import to_categorical
import pickle

params = {
    'dim' : 300,            # dimension of embeddings
    'maximum_steps' : 1000, # number of training steps        
    'lstm_size' : 150,      # dimension of LSTM
    'batch_size' : 25,      # batch size
    'max_words' : 10000,    # maximum number of words to embed
    'padding_size' : 20,    # maximum sentence size
    'num_classes' : 14,     # number of unique classes
    'save_dir' : 'models/' # directory to save hash tables, model weights, etc.
}


# In[238]:


def save_obj(directory, obj, name):
    '''Helper function using pickle to save and load objects'''
    with open(directory + name + '.pkl', 'wb+') as f:
        pickle.dump(obj, f, pickle.HIGHEST_PROTOCOL)

def load_obj(directory, name):
    '''Helper function using pickle to save and load objects'''
    with open(directory + name + ".pkl", "rb") as f:
        return pickle.load(f)
   
def load_data(file = "data/train.txt"):
    '''Helper function to load and transform inputs and labels
    included as a separate function due to NER-specific evaluation needs:
        tensorflow does not have multi-class precision/accuracy as a metric
        so data_y is needed to manually calculate evaluations'''
    file = open(file, 'r')
    sentence, labels = [], []
    data_x, data_y = [], []
    for line in file:
        line = line.strip("\n").split("\t")
       
        # lines with len > 1 are words
        if len(line) > 1:
            sentence.append(line[1])
            labels.append(line[0][2:]) if len(line[0]) > 1 else labels.append(line[0])
       
        # lins with len == 1 are sentence breaks
        if len(line) == 1:
            data_x.append(' '.join(sentence))
            data_y.append(labels)
            sentence, labels = [], []
    return data_x, data_y

def make_tokenizer(file = "data/train.txt", params = params):
    ''' In order for one hot encoding of words and labels to work,
    every word and label has to be seen at least once to make a hashing table.
    This function outputs hash tables for the words and the labels
    that can be used to one-hot-encode them in the generator
    '''
    # Load parameters and data
    max_words = params['max_words']
    padding_size = params['padding_size']
    save_dir = params['save_dir']
    data_x, data_y = load_data(file)
           
    # Use the Keras tokenizer API to generate hashing table for data_x
    tokenizer = Tokenizer(num_words = max_words)
   
    tokenizer.fit_on_texts(data_x)
    word_index = tokenizer.word_index
   
    # Flatten data_y and create hashing table using set logic
    data_y_flattened = [item for sublist in data_y for item in sublist]
    data_x_flattened = [item for sublist in data_x for item in sublist]
   
    labels_index = dict([(y, x + 1) for x, y in enumerate(sorted(set(data_y_flattened)))])
    labels = []
    for item in data_y:
        labels.append([labels_index.get(i) for i in item])
    labels_lookup = {v : k for k, v in labels_index.items()} # reverse dictionary for lookup
    # save hash tables to disk for model serving
    for item, name in zip([word_index, labels_index, labels_lookup],
                          ["word_index", "labels_index", "labels_lookup"]):
        save_obj(save_dir, item, name)
    return word_index, labels_index, labels_lookup

word_index, labels_index, labels_lookup = make_tokenizer()


# In[239]:


word_index


# In[240]:


labels_lookup


# In[241]:


labels_index


# In[242]:


def generate_batches(file = "data/train.txt", params = params, train = True):
    ''' Generate minibatch with dimensions:
    batch_x : (batch_size, max_len)
    lengths : (batch_size,)
    batch_y : (batch_size, num_classes)
   
    file : path to .txt containing training data in BIO format
    '''
   
    batch_size = params['batch_size']
    max_len = params['padding_size']
    save_dir = params['save_dir']
   
    # load hash tables for tokenization
    for item, name in zip([word_index, labels_index, labels_lookup],
                          ["word_index", "labels_index", "labels_lookup"]):
        item = load_obj(save_dir, name)
   
    while True:
        with open(file, 'r') as f:
            batch_x, lengths, batch_y = [], [], []
            words, labels = [], []
            for line in f:
                line = line.strip("\n").split("\t")
                # lines with len > 1 are words
                if len(line) > 1:
                    labels.append(line[0][2:]) if len(line[0]) > 1 else labels.append(line[0])
                    words.append(line[1])

                # lines with len == 1 are breaks between sentences
                if len(line) == 1:
                    words = [word_index.get(x) if x in word_index.keys() else 0 for x in words]
                    labels = [labels_index.get(y) for y in labels]
                    batch_x.append(words)
                    batch_y.append(labels)
                    lengths.append(min(len(words), max_len))
                    words, labels = [], []

                if len(batch_x) == batch_size:
                    batch_x = pad_sequences(batch_x, maxlen = max_len, value = 0, padding = "post")
                    batch_y = pad_sequences(batch_y, maxlen = max_len, value = 0, padding = "post")
                    yield (batch_x, lengths), batch_y
                    batch_x, lengths, batch_y = [], [], []
            if train == False:
                break


# In[243]:


# For model training, we need an input function that will feed a tf.Dataset
def input_fn(file, params = None, train = True):
    params = params if params is not None else {}
    shapes = (([None, None], [None]), [None, None]) # batch_x, lengths, batch_y shapes
    types = ((tf.int32, tf.int32), tf.int32)        # batch_x, lengths, batch_y data types
   
    generator = partial(generate_batches, file, train = train)
    dataset = tf.data.Dataset.from_generator(generator, types, shapes)
    return dataset


# For model serving, we need a serving function that will feed tf.placeholders
def serving_input_fn():
    words = tf.placeholder(dtype=tf.int32, shape=[None, None], name='words')
    length = tf.placeholder(dtype=tf.int32, shape=[None], name='length')
    receiver_tensors = {'words': words, 'length': length}
    features = {'words': words, 'length': length}
    return tf.estimator.export.ServingInputReceiver(features, receiver_tensors)


# In[ ]:





# In[244]:


#import keras
#import keras.utils
#from keras import utils as np_utils

import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

print(tf.__version__)
print(keras.__version__)


# In[245]:




def model_fn(features, labels, mode, params = params):
    # import the data and unpack the features
    # serving input_fn returns a dict, convert to multivalue obj
    if isinstance(features, dict):
        features = features['words'], features['length']
   
    words, length = features
   
    # Embedding
    embedding = tf.Variable(tf.random.normal([params['max_words'], params['dim']]))
    embedding_lookup_for_x = tf.nn.embedding_lookup(embedding, words)
   
    # LSTM
    lstm_cell_fw = tf.keras.layers.LSTMCell(params['lstm_size'])
    lstm_cell_bw = tf.keras.layers.LSTMCell(params['lstm_size'])
    states, final_state = tf.keras.layers.Bidirectional(
                                        cell_fw = lstm_cell_fw,
                                        cell_bw = lstm_cell_bw,
                                        inputs = embedding_lookup_for_x,
                                        dtype = tf.float32,
                                        time_major = False,
                                        sequence_length = length)
    lstm_out = tf.keras.layers.concat([states[0], states[1]], axis = 2)

    

#tf.compat.v1.keras.layers.LSTMCell
#keras.layers.Bidirectional(keras.layers.RNN(cell))
    
    
    # LSTM
#    lstm_cell_fw = tf.nn.rnn_cell.BasicLSTMCell(params['lstm_size'], state_is_tuple = True)
#    lstm_cell_bw = tf.nn.rnn_cell.BasicLSTMCell(params['lstm_size'], state_is_tuple = True)
#    states, final_state = tf.nn.bidirectional_dynamic_rnn(
#                                        cell_fw = lstm_cell_fw,
#                                        cell_bw = lstm_cell_bw,
#                                        inputs = embedding_lookup_for_x,
#                                        dtype = tf.float32,
#                                        time_major = False,
#                                        sequence_length = length)
#    lstm_out = tf.concat([states[0], states[1]], axis = 2)
    
       
    # Conditional random fields
    logits = tf.layers.dense(lstm_out, params['num_classes'])
    crf_params = tf.get_variable("crf", [params['num_classes'], params['num_classes']],
                                 dtype=tf.float32)
    pred_ids, _ = tf.contrib.crf.crf_decode(logits, crf_params, length)
    training = (mode == tf.estimator.ModeKeys.TRAIN)
   
    # Prediction
    if mode == tf.estimator.ModeKeys.PREDICT:
        predictions = {
            'pred_ids': pred_ids,
            'tags': words,
            'length' : length,
        }
        export_outputs = {
          'prediction': tf.estimator.export.PredictOutput(predictions)
      }
       
        return tf.estimator.EstimatorSpec(mode, predictions=predictions,
                                          export_outputs=export_outputs)
   
    # Loss functions and optimizers
    log_likelihood, _ = tf.contrib.crf.crf_log_likelihood(
        logits, labels, length, crf_params)
   
    loss = tf.reduce_mean(-log_likelihood)
    train_op = tf.train.AdamOptimizer().minimize(
        loss, global_step = tf.train.get_or_create_global_step())
       
    # Training
    if mode == tf.estimator.ModeKeys.TRAIN:
        return tf.estimator.EstimatorSpec(mode = mode,
                                           loss = loss,
                                           train_op = train_op)


# In[246]:


# Spin up the estimator
config = tf.estimator.RunConfig()
estimator = tf.estimator.Estimator(model_fn, 'models/model', config, params)

# Create train spec
train_input_fn = partial(input_fn, "data/train.txt", params = params)
train_spec = tf.estimator.TrainSpec(train_input_fn)

# Create evaluation spec
eval_input_fn = partial(input_fn, "data/test.txt", params = params, train = False)
eval_spec = tf.estimator.EvalSpec(eval_input_fn)


# In[247]:


import time
ts = time.time()
estimator.train(input_fn = train_input_fn, max_steps = 1000)
te = time.time()
print("Completed in {} seconds".format(int(te - ts)))
estimator.export_savedmodel('models/saved_model/', serving_input_fn)


# In[130]:


# Generate predictions
predictions = estimator.predict(eval_input_fn)

# Load hash tables and true labels
labels_index = load_obj(params['save_dir'], "labels_index")
_, true = load_data("data/test.txt")

# Specify which label_index is non-entity
dummy_label = labels_index.get("O")

# Convert [[string, string], [string, string] ...] to [[int, int], [int, int]]
# with hashing table for label indexes
labels = []
for row in true:
    labels.append([labels_index.get(y) for y in row])
   
# Loop through preds, labels and calculate metrics
precisions, recalls, f1s = [], [], []
for pred, true in zip(predictions, labels):
    pred = pred['pred_ids'][:pred['length']] # undo pad_sequences
    pred = [x for x in pred if x != dummy_label] # remove preds that aren't entities
    true = np.asarray([x for x in true if x != dummy_label])
    recall = calc_recall(true, pred)
    recalls.append(recall)
    precision = calc_precision(true, pred)
    precisions.append(precision)
    f1s.append(calc_f1(precision, recall))
   
print("Precision: {} \nRecall: {} \nF1-score: {}".format(np.around(np.mean(precisions), 3),
                                                         np.around(np.mean(recalls), 3),
                                                         np.around(np.mean(f1s), 3)))


# In[211]:


from pathlib import Path
from tensorflow.contrib import predictor

LINE = 'did george clooney make a science fiction movie in the 1980s'


def predict(line, export_dir = 'models/saved_model/', params = params):
    # Load hash tables
    word_index = load_obj(params['save_dir'], "word_index")
    labels_lookup = load_obj(params['save_dir'], "labels_lookup")
   
    # Identify and load model weights
    subdirs = [x for x in Path(export_dir).iterdir()
                   if x.is_dir() and 'temp' not in str(x)]
    latest_model = str(sorted(subdirs)[-1])
    predict_fn = predictor.from_saved_model(latest_model)
               
    # Preprocess sentence input
    line = line.strip().split()
    vector = [word_index.get(x) if x in word_index.keys() else 0 for x in line] # tokenize
    vector[len(vector):20] = [0] * (20 - len(vector)) # pad prediction
       
    # Calculate precision and transform for display
    predictions = predict_fn({'words': [vector], 'length': [len(line)]})
    tags = predictions.get('tags')
    preds = predictions.get('pred_ids')
    for tag, pred in zip(tags, preds):
        tag = [word for word in tag if word != 0] # unpad
        pred = pred[:len(tag)]
        pred = [labels_lookup.get(num) for num in pred] #untokenize
        print(line, "\n", pred)
   
predict(LINE)


# In[ ]:





# In[ ]:

**Comment by Saduf2019** [comment]
@bavanhassel 
Please share the code as a colab gist [google colab gist], indentation errors in this code will delay the replication of issue faced.

**Comment by bavanhassel** [comment]
https://colab.research.google.com/gist/bavanhassel/01f4fddf93ca028671cf7f280018e54c/named_entity_recognition.ipynb

**Comment by bavanhassel** [comment]
https://colab.research.google.com/gist/bavanhassel/9efd30c15d02f64a5454e45ae0dfc665/try_keras_ner_github.ipynb

**Comment by Saduf2019** [comment]
@bavanhassel 
Please share "ner_dataset.csv" for us to replicate the issue on our local.

**Comment by bavanhassel** [comment]
https://www.kaggle.com/abhinavwalia95/loading-named-entity-recognition-training-dataset

**Comment by Saduf2019** [comment]
@bavanhassel 
Please provide the csv directly, i cannot access kaggle and download it.

**Comment by bavanhassel** [comment]
This application does not support uploading csv files

**Comment by bavanhassel** [comment]
[ner_dataset.txt](https://github.com/tensorflow/tensorflow/files/5333738/ner_dataset.txt)
You can change the file extension back to csv

**Comment by Saduf2019** [comment]
@bavanhassel 
I ran the code shared and face a different error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/4174830821f613f3c07332258f5069a4/untitled425.ipynb).

**Comment by bavanhassel** [comment]
[test.txt](https://github.com/tensorflow/tensorflow/files/5335251/test.txt)
[train.txt](https://github.com/tensorflow/tensorflow/files/5335252/train.txt)

**Comment by Saduf2019** [comment]
@bavanhassel
I ran the code shared and face a different error, please find the [gist here.](https://colab.research.google.com/gist/Saduf2019/cad9490be5a452e5bfbf77b6a28906dc/untitled429.ipynb).
It is not possible to replicate the issue faced unless all dependencies are shared.
